src/common/viz/FigureExtractor.js

```diff
@@ -0,0 +1,145 @@
+        let desc = {
+            id: id,
+            execId: execId,
+            type: 'graph',
+            name: node.getAttribute('name'),
+            graphId: node.getAttribute('id'),
+            title: node.getAttribute('title'),
+        };
+
+        let childrenIds = node.getChildrenIds();
+        let childNode, childNodeFn;
+        desc.subGraphs = childrenIds.map((childId) => {
+            childNode = this._client.getNode(childId);
+            childNodeFn = this.getMetaType(childNode);
+            return this[childNodeFn](childNode);
+        });
+        desc.subGraphs.sort(this.compareSubgraphIDs);
+        return desc;
+    };
+
+    FigureExtractor.prototype[EXTRACTORS.SUB_GRAPH] = function (node) {
+        const [id, graphId, jobId, execId] = this.getIdHierarchy(node, 4); // eslint-disable-line no-unused-vars
+        let desc;
+
+        desc = {
+            id: id,
+            execId: execId,
+            type: 'subgraph',
+            graphId: this._client.getNode(graphId).getAttribute('id'),
+            subgraphId: node.getAttribute('id'),
+            subgraphName: node.getAttribute('name'),
+            title: node.getAttribute('title'),
+            xlim: node.getAttribute('xlim'),
+            ylim: node.getAttribute('ylim'),
+            xlabel: node.getAttribute('xlabel'),
+            ylabel: node.getAttribute('ylabel'),
+        };
+
+        const children = node.getChildrenIds().map(id => this._client.getNode(id));
+        desc.lines = children.filter(node => this.getMetaType(node) === EXTRACTORS.LINE)
+            .map(lineNode => this[this.getMetaType(lineNode)](lineNode));
+        desc.images = children.filter(node => this.getMetaType(node) === EXTRACTORS.IMAGE)
+            .map(imageNode => this[this.getMetaType(imageNode)](imageNode));
+
+        return desc;
+    };
+
+    FigureExtractor.prototype[EXTRACTORS.LINE] = function (node) {
+        const ids= this.getIdHierarchy(node, 5),
+            id = ids[0],
+            execId = ids[ids.length-1];
+        let points, desc;
+
+        points = node.getAttribute('points').split(';')
+            .filter(data => !!data)  // remove any ''
+            .map(pair => {
+                const [x, y] = pair.split(',').map(num => parseFloat(num));
+                return {x, y};
+            });
+        desc = {
+            id: id,
+            execId: execId,
+            subgraphId: this._client.getNode(node.getParentId()).getAttribute('id'),
+            lineName: node.getAttribute('name'),
+            label: node.getAttribute('label'),
+            name: node.getAttribute('name'),
+            type: 'line',
+            points: points,
+            color: node.getAttribute('color')
+        };
+
+        return desc;
+    };
+
+    FigureExtractor.prototype[EXTRACTORS.IMAGE] = function (node) {
+        const ids= this.getIdHierarchy(node, 5),
+            id = ids[0],
+            execId = ids[ids.length-1],
```

Comment:
Do we always know that the execution is the 4th node up in the containment hierarchy? It might be better to get the ID of the first ancestor of the given type... Since it looks like this is mostly needed for the execution, you could just have a method like `getExecutionId(node)` where it finds the first parent that is an execution and returns the ID.